flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/sink/TiDBDataStreamSinkProvider.java

```diff
@@ -0,0 +1,274 @@
+
+    // add RowConvertMapFunction
+    TiDBRowConverter tiDBRowConverter = new TiDBRowConverter(tiTableInfo);
+    DataStream<Row> tiRowDataStream = dataStream.map(new RowConvertMapFunction(tiDBRowConverter,
+        sinkOptions.isIgnoreAutoincrementColumn()));
+
+    if (sinkTransaction == MINIBATCH) {
+      LOG.info("Flink sinkTransaction is working on mode miniBatch");
+
+      // mini batch use row buffer deduplicate
+      TiDBWriteOperator tiDBWriteOperator = new TiDBMiniBatchWriteOperator(databaseName, tableName, properties,
+          timestamp, sinkOptions, createRowIdStarts(clientSession, parallelism));
+      SingleOutputStreamOperator<Void> transform = tiRowDataStream.transform("PRE_WRITE",
+          Types.VOID,
+          tiDBWriteOperator);
+
+      // Since sink happens in operator, all we need is to discard stream.
+      return transform
+          .addSink(new DiscardingSink<>())
+          .setParallelism(1)
+          .name(DiscardingSink.class.getSimpleName());
+
+    } else if (sinkTransaction == CHECKPOINT) {
+      LOG.info("Flink sinkTransaction is working on mode checkpoint");
+
+      // validate if CheckpointingEnabled.
+      CheckpointConfig checkpointConfig = dataStream.getExecutionEnvironment()
+          .getCheckpointConfig();
+      if (!checkpointConfig.isCheckpointingEnabled()) {
+        throw new IllegalStateException(
+            "Checkpoint transaction is invalid for stream without checkpoint");
+      }
+
+      tiRowDataStream = deduplicate(tiRowDataStream, tiTableInfo);
+
+      TiDBSinkFunction sinkFunction = new TiDBSinkFunction(
+          new TiDBTransactionStateSerializer(),
+          new TiDBTransactionContextSerializer(),
+          databaseName,
+          tableName,
+          properties,
+          sinkOptions);
+      return tiRowDataStream.addSink(sinkFunction);
+
+    } else if (sinkTransaction == GLOBAL) {
+      LOG.info("Flink sinkTransaction is working on mode global");
+
+      if (!context.isBounded()) {
+        throw new IllegalStateException(
+            "Global transaction is invalid for streaming mode or unbounded stream");
+      }
+
+      tiRowDataStream = deduplicate(tiRowDataStream, tiTableInfo);
+
+      // preWrite primary key, use random uuid as primary key.
+      TiDBWriteHelper tiDBWriteHelper = new TiDBWriteHelper(clientSession.getTiSession(),
+          timestamp.getVersion());
+      tiDBWriteHelper.preWriteFirst(new BytePairWrapper(fakePrimaryKey(), new byte[0]));
+      primaryKey = tiDBWriteHelper.getPrimaryKeyMust();
+      tiDBWriteHelper.close();
+
+      // add operator which preWrite secondary keys.
+      TiDBWriteOperator tiDBWriteOperator = new TiDBGlobalWriteOperator(databaseName, tableName, properties,
+          timestamp, sinkOptions, primaryKey, createRowIdStarts(clientSession, parallelism));
+      SingleOutputStreamOperator<Void> transform = tiRowDataStream.transform("PRE_WRITE",
```

Comment:
should we setParallelism in this operator

Reply:
nope, it's decided by user